std/typecons.d

```diff
@@ -149,25 +157,37 @@ public:
     ~this()
     {
         debug(Unique) writeln("Unique destructor of ", (_p is null)? null: _p);
-        if (_p !is null) delete _p;
-        _p = null;
+        release();
     }
-    /** Returns whether the resource exists. */
-    @property bool isEmpty() const
-    {
-        return _p is null;
-    }
-    /** Transfer ownership to a $(D Unique) rvalue. Nullifies the current contents. */
-    Unique release()
+
+    /** Frees the underlying $(D RefT) and nulls it */
+    void release()
     {
         debug(Unique) writeln("Release");
-        auto u = Unique(_p);
-        assert(_p is null);
-        debug(Unique) writeln("return from Release");
-        return u;
+        if (_p !is null)
+        {
+            import core.memory : GC;
+
+            destroy(_p);
+            GC.free(cast(void*)_p);
+            _p = null;
+        }
     }
-    /** Forwards member access to contents. */
-    RefT opDot() { return _p; }
+
+    /**
+    Returns the underlying $(D RefT) for use by non-owning code.
+
+    The holder of a $(D Unique!T) is the <em>owner</em> of that $(D T).
+    For code that does not own the resource (and therefore does not affect
+    its life cycle), pass a plain old reference.
+    */
+    RefT get() { return _p; }
+
+    /**
+    Allows you to dereference the underlying $(D RefT)
+    and treat it like a $(D RefT) in other ways (such as comparing to null)
+    */
+    alias _p this;
```

Comment:
You're right, they can, and perhaps I should add a comment saying as much. However, this is not inherently bad. `Unique` conveys _ownership_, and non-owning functions shouldn't pass around smart pointers if they aren't going to play a role in the underlying object's lifetime.

See [this bit (starting at the 14 minute mark or so)](https://youtu.be/xnqTKD8uD64?t=14m31s) of a recent Herb Sutter talk. Yes, it's about C++, but the exact same semantics are at play here and he discusses it extensively. He makes these points much more eloquently than I do, so I hope you take the time to watch it, but to summarize, doing this bakes the ownership semantics of calls directly into the function signatues. Observe:

```d
void consume(Unique!T u) // Takes ownership of u

void poke(ref T u) // Uses u but plays no impact on its lifetime

void opt(T* u) // u is optional; has no impact on its lifetime
```